src/_backend_agg.h

```diff
@@ -734,41 +734,68 @@ inline void RendererAgg::draw_text_image(GCAgg &gc, ImageArray &image, int x, in
+        rect.line_to(image.dim(1), 0);
+        rect.line_to(image.dim(1), image.dim(0));
+        rect.line_to(0, image.dim(0));
+        rect.line_to(0, 0);
+        agg::conv_transform<agg::path_storage> rect2(rect, mtx);
+
+        agg::trans_affine inv_mtx(mtx);
+        inv_mtx.invert();
+
+        agg::image_filter_lut filter;
+        filter.calculate(agg::image_filter_spline36());
+        interpolator_type interpolator(inv_mtx);
+        color_span_alloc_type sa;
+        image_accessor_type ia(pixf_img, agg::gray8(0));
+        image_span_gen_type image_span_generator(ia, interpolator, filter);
+        span_gen_type output_span_generator(&image_span_generator, gc.color);
+        renderer_type ri(rendererBase, sa, output_span_generator);
+
+        theRasterizer.add_path(rect2);
+        agg::render_scanlines(theRasterizer, slineP8, ri);
+    } else {
+        agg::rect_i fig, text;
+
+        fig.init(0, 0, width, height);
+        text.init(x, y - image.dim(0), x + image.dim(1), y);
+        text.clip(fig);
+
+        if (gc.cliprect.x1 != 0.0 || gc.cliprect.y1 != 0.0 || gc.cliprect.x2 != 0.0 || gc.cliprect.y2 != 0.0) {
+            agg::rect_i clip;
+
+            clip.init(int(mpl_round(gc.cliprect.x1)),
+                      int(mpl_round(gc.cliprect.y1)),
+                      int(mpl_round(gc.cliprect.x2)),
+                      int(mpl_round(gc.cliprect.y2)));
+            text.clip(clip);
+        }
 
-    agg::rendering_buffer srcbuf(
-        image.data(), (unsigned)image.dim(1), (unsigned)image.dim(0), (unsigned)image.dim(1));
-    agg::pixfmt_gray8 pixf_img(srcbuf);
-
-    agg::trans_affine mtx;
-    mtx *= agg::trans_affine_translation(0, -image.dim(0));
-    mtx *= agg::trans_affine_rotation(-angle * agg::pi / 180.0);
-    mtx *= agg::trans_affine_translation(x, y);
-
-    agg::path_storage rect;
-    rect.move_to(0, 0);
-    rect.line_to(image.dim(1), 0);
-    rect.line_to(image.dim(1), image.dim(0));
-    rect.line_to(0, image.dim(0));
-    rect.line_to(0, 0);
-    agg::conv_transform<agg::path_storage> rect2(rect, mtx);
-
-    agg::trans_affine inv_mtx(mtx);
-    inv_mtx.invert();
-
-    agg::image_filter_lut filter;
-    filter.calculate(agg::image_filter_spline36());
-    interpolator_type interpolator(inv_mtx);
-    color_span_alloc_type sa;
-    image_accessor_type ia(pixf_img, agg::gray8(0));
-    image_span_gen_type image_span_generator(ia, interpolator, filter);
-    span_gen_type output_span_generator(&image_span_generator, gc.color);
-    renderer_type ri(rendererBase, sa, output_span_generator);
-
-    theRasterizer.add_path(rect2);
-    agg::render_scanlines(theRasterizer, slineP8, ri);
+        for (int yi = text.y1; yi < text.y2; ++yi) {
+            for (int xi = text.x1; xi < text.x2; ++xi) {
+                typename ImageArray::value_type pixel = image(
+                        yi - (y - image.dim(0)), xi - x);
+                pixFmt.blend_pixel(xi, yi, gc.color, pixel);
```

Comment:
I'm not an agg expert, but it seems that blend_color_hspan could be even faster.

http://sourceforge.net/p/vector-agg/mailman/message/5116797/